File: src/docforge/title.py
def extract_title(
    h1_text: str | None,
    soup_title_text: str | None,
    fallback_stem: str,
) -> str:
    """Resolve title: body h1 → HTML <title> → filename stem.

    Empty / whitespace-only inputs fall through to the next tier.
    """
    if h1_text and h1_text.strip():
        return h1_text.strip()
    if soup_title_text and soup_title_text.strip():
        return soup_title_text.strip()
    return fallback_stem

File: src/title.ts
export function extractTitle(
  h1Text: string | null,
  soupTitleText: string | null,
  fallbackStem: string,
): string {
  if (h1Text && h1Text.trim()) return h1Text.trim();
  if (soupTitleText && soupTitleText.trim()) return soupTitleText.trim();
  return fallbackStem;
}
